bite-size-notes/week-6-b-free-final-time-PMP.md

```diff
@@ -0,0 +1,231 @@
+
+### Pontryagin's Minimum Principle (PMP)
+
+Let the cost functional be
+$$\underbrace{J[z(t),u(t),t]}_{\text{cost functional}} = \int_0^T \underbrace{L\big(z(t), u(t), t\big)}_{\text{cost-to-go}} \,dt +  \underbrace{\Psi(T)}_{\text{terminal cost}}$$
+as defined above.
+
+The constraints from the system dynamics are added to the Lagrangian $L$ with time-varying Lagrange multipliers $\lambda(t)$  whose elements are called the co-states of the system.
+
+We can then construct the Hamiltonian $H$:
+$$
+\begin{equation*}
+H\big(z(t), u(t), \lambda(t), t\big) = \lambda(t)^\top \cdot \dot{z} + L\big(z(t), u(t)\big).
+\end{equation*}
+$$
+
+Pontryagin's minimum principle states that the optimal parameters must minimize the Hamiltonian $H$ for all time $t \in [0, T]$ and all feasible controls $u$ so that
+$$
+\begin{equation}
+H(z^*(t), u^*(t), \lambda^*(t), t) \leq H(z(t), u(t), \lambda(t), t).
+\end{equation}
+$$
+
+$$
+\begin{equation}
+\begin{aligned}
+-\dot{\lambda}(t)
+&= \nabla_z H(z^*(t), u^*(t), \lambda^*(t), t) \\ 
+&= \lambda(t)^\top \cdot \nabla_z \dot{z}(z^*(t), u^*(t)) + \nabla_z L(z^*(t), u^*(t)) 
+\end{aligned}
+\end{equation}
+$$
+
+$$
+\begin{equation}
+\lambda(T)^\top = \nabla_z \Psi(z(T))
+\end{equation}
+$$
+
+The above equations are necessary for optimality with fixed final time, this last condition is for free final time.
+$$
+\begin{equation}
+\nabla_T \Psi(z(T)) + H(T) = 0
+\end{equation}
+$$
+
+### Transversality Conditions
+
+For free final time, the transversality conditions state that at the optimal terminal time $T$, the Hamiltonian $H$ should satisfy
+$$H\big(z(T),u(T),\lambda(T),T\big) = 0.$$ This implies that no further performance can be achieved by changing $T$.
+
+For similar reasons, all co-states should be zero at the optimal terminal time $T$.
+$$\lambda(T) = 0$$
+
+
+## Example: 2D Unicycle
+
+Now let's look at a standard optimal control problem, but this time with terminal time being an optimization variable.
+
+### Analytical solution
+Let's revisit a common example: the 2D unicycle. We seek to reach a goal state and minimize both effort and time doing so.
+* state variable $z$
+  * positions $x$, $y$
+  * velocity $v$
+  * angle $\theta$
+* control input $u$
+  * acceleration $a$
+  * angular velocity $\omega$
+* state variable $z_g$
+  * goal positions $x_g$, $y_g$
+  * goal velocity $v_g$
+  * goal angle $\theta_g$
+* co-state $\lambda$
+* Lagrange multiplier $\mu$
+* maximum acceleration $a_{\text{max}}$
+
+Then the state and dynamics vectors, still functions of time, are:
+
+$$
+z = \begin{bmatrix}
+  x \\
+  y \\
+  v \\
+  \theta
+\end{bmatrix}, \qquad
+\text{and}
+\qquad \dot{z} = \begin{bmatrix}
+  v \cos \theta \\
+  v \sin \theta \\
+  a \\
+  \omega
+\end{bmatrix}
+$$
+
+Let the Lagrangian $L(z(t), u(t), t\big) = a^2 + \omega^2 + \|x-x_g\|^2 + \|y-y_g\|^2$ and the terminal cost be $\Psi(T) = \|z(T) - z_g\|^2$. Include a constraint on the control: $a \le a_{\text{max}}$.
```

Comment:
deceleration magnitude is not handled with this constraint. Is this the case? If yes please clarify this point; if not, please change the constraint accordingly

Reply:
Correct, that was an oversight on my end, I changed the constraint to be a 2-norm (also makes it convex).